src/planscape/forsys/cluster_stands.py

```diff
@@ -0,0 +1,261 @@
+                (num_connected_components) +
+                " it's impossible to convert this into %d clusters" %
+                (num_clusters))
+
+        self._normalized_pixel_dist_to_condition_values = \
+            self._normalize_condition_values(
+                pixel_dist_to_condition_values, priority_condition_max_value)
+        self._normalized_priority_weights = self._normalize_priority_weights(
+            priority_weights)
+
+        self._features = self._get_features(
+            self._normalized_pixel_dist_to_condition_values, pixel_width,
+            pixel_height, self._normalized_priority_weights, pixel_index_weight)
+
+        ward = AgglomerativeClustering(
+            n_clusters=num_clusters, linkage='ward',
+            connectivity=self._connectivity).fit(
+            np.array(self._features))
+
+        self.clusters_to_stands = self._get_cluster_pixels(
+            ward.labels_, self._mask, pixel_width, pixel_height)
+
+    def _validate_input_params(
+            self,
+            pixel_dist_to_condition_values: dict[int,
+                                                 dict[int, dict[str, float]]],
+            pixel_width: int, pixel_height: int,
+            priority_condition_max_value: float,
+            priority_weights: dict[str, float],
+            pixel_index_weight: float, num_clusters: int) -> None:
+        if pixel_width <= 0 or type(pixel_height) != int:
+            raise Exception("expected pixel_width to be a positive integer")
+        if pixel_height <= 0 or type(pixel_height) != int:
+            raise Exception("expected pixel_height to be a positive integer")
+        if num_clusters < 1:
+            raise Exception("num_clusters must be a positive integer")
+        if pixel_index_weight < 0:
+            raise Exception("pixel_index_weight must be gte 0")
+        priorities = list(priority_weights.keys())
+        if len(priorities) == 0:
+            raise Exception("expected at least 1 priority weight")
+        for x in pixel_dist_to_condition_values.keys():
+            for y in pixel_dist_to_condition_values[x].keys():
+                conditions = pixel_dist_to_condition_values[x][y].keys()
+                if len(priorities) != len(conditions):
+                    raise Exception(
+                        "expected len(priorities) == len(conditions)")
+                for p in priorities:
+                    if p not in conditions:
+                        raise Exception(
+                            "expected conditions to include priority, %s" %
+                            (p))
+                    v = pixel_dist_to_condition_values[x][y][p]
+                    if v < 0 or v > priority_condition_max_value:
+                        raise Exception("expected condition score to be " +
+                                        "within range, [0, %f]" % (
+                                            priority_condition_max_value))
```

Comment:
should we be using the same max value for all types of conditions?

Reply:
also, condition scores can be negative, are we handling that before we call this?

Reply:
yes: we expect normalized conditions to have the same max value. I've added this as a comment.

normalized condition scores can be negative, but I don't think the logic to convert [-1, 1] to [0, 2] is within the scope of the clusterer. #597 shows where in the flow the conversion takes place.